feat: add HMR plugin (#209)
sapphiredev/plugins: packages/hmr/README.md

[![npm](https://img.shields.io/npm/v/@sapphire/plugin-hmr?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/@sapphire/plugin-hmr)

</div>

## Description

This plugin provides a HMR (Hot Module Replacement) feature for @sapphire/framework. This allows you to add, delete, and
update commands, listeners, and more without having to restart your bot. This allows for rapid iteration and prototyping
without your bot slowing you down.

## Features

- Fully ready for TypeScript!
- Includes ESM ready entrypoint
- FAST!
- Super easy to use

## Installation

`@sapphire/plugin-hmr` depends on the following packages. Be sure to install these along with this package!

- [`@sapphire/framework`](https://www.npmjs.com/package/@sapphire/framework)

You can use the following command to install this package, or replace `npm install` with your package manager of choice.

```sh
npm install @sapphire/plugin-hmr @sapphire/framework
```

---

## Usage

### JavaScript

in your main or setup file, register the plugin:

```javascript
require('@sapphire/plugin-hmr/register');
```

if you want to make sure its only loaded in development, you can do this:

```typescript

> if (process.env.NODE_ENV === 'development') import('@sapphire/plugin-hmr/register');

Reviewer: Should include the docs for `start()` as exported from the index.

```

### TypeScript

in your main or setup file, register the plugin:

```typescript

> import('@sapphire/plugin-hmr/register');

Reviewer: ```suggestion
import '@sapphire/plugin-hmr/register';
```

```

if you want to make sure its only loaded in development, you can do this:

```typescript

> if (process.env.NODE_ENV === 'development') import('@sapphire/plugin-hmr/register');

Reviewer: Ditto.

</div>

## Description

This plugin provides a HMR (Hot Module Replacement) feature for @sapphire/framework. This allows you to add, delete, and
update commands, listeners, and more without having to restart your bot. This allows for rapid iteration and prototyping
without your bot slowing you down.

## Features

- Fully ready for TypeScript!
- Includes ESM ready entrypoint
- FAST!
- Super easy to use

## Installation

`@sapphire/plugin-hmr` depends on the following packages. Be sure to install these along with this package!

- [`@sapphire/framework`](https://www.npmjs.com/package/@sapphire/framework)

You can use the following command to install this package, or replace `npm install` with your package manager of choice.

```sh
npm install @sapphire/plugin-hmr @sapphire/framework
```

---

## Usage

### JavaScript

in your main or setup file, register the plugin:

```javascript
require('@sapphire/plugin-hmr/register');
```

if you want to make sure its only loaded in development, you can do this:

```javascript
const { start } = require('@sapphire/hmr');

> if (process.env.NODE_ENV === 'development') start();

Reviewer: Clarify with a comment that this needs to be called after `SapphireClient#login`, so it registers the hooks for all the stores and all their paths (user paths are registered in login).

Perhaps something like:

```javascript
// Define client and everything

async function main() {
	await client.login();

	if (process.env.NODE_ENV === 'development') start();
}

void main();
```

```

### TypeScript

in your main or setup file, register the plugin:

```typescript
import '@sapphire/plugin-hmr/register';
```

if you want to make sure its only loaded in development, you can do this:

```typescript

> // at the top of your file
> import { start } from '@sapphire/plugin-hmr';

Reviewer: Why does this have a comment while the other doesn't? And it's unneeded.

```suggestion
import { start } from '@sapphire/plugin-hmr';
```

## Features

-   Fully ready for TypeScript!
-   Includes ESM ready entrypoint
-   Easy to use

## Installation

`@sapphire/plugin-hmr` depends on the following packages. Be sure to install these along with this package!

-   [`@sapphire/framework`](https://www.npmjs.com/package/@sapphire/framework)

You can use the following command to install this package, or replace `npm install` with your package manager of choice.

```sh
npm install @sapphire/plugin-hmr @sapphire/framework
```

---

## Usage

### JavaScript

In your main or setup file, register the plugin:

```javascript
require('@sapphire/plugin-hmr/register');
```

Or if you want to make sure the plugin is only loaded in development, you can register it dynamically like so:

```javascript
const { start } = require('@sapphire/hmr');

const client = new SapphireClient({
	/* your bot options */
});

async function main() {
	await client.login();

	// this has to be called after you have logged into your bot
	if (process.env.NODE_ENV === 'development') {
		start();
	}
}

> main();

Reviewer: ```suggestion
void main();
```

<div align="center">

![Sapphire Logo](https://cdn.skyra.pw/gh-assets/sapphire-banner.png)

# @sapphire/plugin-hmr

**Plugin for <a href="https://github.com/sapphiredev/framework">@sapphire/framework</a> for super-speed HMR.**

[![GitHub](https://img.shields.io/github/license/sapphiredev/plugins)](https://github.com/sapphiredev/plugins/blob/main/LICENSE.md)
[![codecov](https://codecov.io/gh/sapphiredev/plugins/branch/main/graph/badge.svg?token=QWL8FB16BR)](https://codecov.io/gh/sapphiredev/plugins)
[![npm bundle size](https://img.shields.io/bundlephobia/min/@sapphire/plugin-hmr?logo=webpack&style=flat-square)](https://bundlephobia.com/result?p=@sapphire/plugin-hmr)
[![npm](https://img.shields.io/npm/v/@sapphire/plugin-hmr?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/@sapphire/plugin-hmr)

</div>

## Description

This plugin provides a HMR (Hot Module Replacement) feature for @sapphire/framework. This allows you to add, delete, and

> update commands, listeners, and more without having to restart your bot. This allows for rapid iteration and prototyping

Reviewer 2: ```suggestion
update commands, listeners, and other pieces without having to restart your bot. This allows for rapid iteration and prototyping
```

without your bot slowing you down.

## Features

> -   Fully ready for TypeScript!
> -   Includes ESM ready entrypoint
> -   FAST!
> -   Super easy to use

Reviewer 2: ```suggestion
-   Fully ready for TypeScript!
-   Includes ESM ready entrypoint
-   Easy to use
```

## Installation

`@sapphire/plugin-hmr` depends on the following packages. Be sure to install these along with this package!

-   [`@sapphire/framework`](https://www.npmjs.com/package/@sapphire/framework)

You can use the following command to install this package, or replace `npm install` with your package manager of choice.

```sh
npm install @sapphire/plugin-hmr @sapphire/framework
```

---

## Usage

### JavaScript

> in your main or setup file, register the plugin:

Reviewer 2: ```suggestion
In your main or setup file, register the plugin:
```

```javascript
require('@sapphire/plugin-hmr/register');
```

> if you want to make sure its only loaded in development, you can do this:

Reviewer 2: ```suggestion
Or if you want to make sure the plugin is only loaded in development, you can register it dynamically like so:
```

```javascript
const { start } = require('@sapphire/hmr');

const client = new SapphireClient({
	/* your bot options */
});

async function main() {
	await client.login();

	// this has to be called after you have logged into your bot

> 	if (process.env.NODE_ENV === 'development') start();

Reviewer 2: ```suggestion
	if (process.env.NODE_ENV === 'development') {
		start();
	}
```

}

main();
```

### TypeScript

> in your main or setup file, register the plugin:

Reviewer 2: ```suggestion
In your main or setup file, register the plugin:
```

```typescript
import '@sapphire/plugin-hmr/register';
```

> if you want to make sure its only loaded in development, you can do this:

Reviewer 2: ```suggestion
Or if you want to make sure the plugin is only loaded in development, you can register it dynamically like so:
```

```typescript
import { start } from '@sapphire/plugin-hmr';

const client = new SapphireClient({
	/* your bot options */
});

async function main() {
	await client.login();

	// this has to be called after you have logged into your bot

> 	if (process.env.NODE_ENV === 'development') start();

Reviewer 2: ```suggestion
	if (process.env.NODE_ENV === 'development') {
		start();
	}
```

}

void main();
```

> In order for proper HMR, you will need to recompile your code, so you will need to edit your package.json:
> 
> - replace the dev and start script 

Reviewer 2: ```suggestion
In order for HMR to pick up your compiled JavaScript files, you will need to recompile your code. To that end, we will configure a `dev` script in `package.json` scripts that runs `build` in parallel with `start`:
```

```json
"scripts": {

> 	"dev": "run-p watch:build start",
> 	"start": "run-s build start:bot"

Reviewer 2: ```suggestion
	"dev": "run-p watch start",
	"build": "tsc",
	"watch": "tsc --watch",
	"start": "node dist/index.js"
```

}
```

> 
> - add a start:bot script
> ```json
> "scripts": {
> 	"start:bot": "node dist/index.js"
> }
> ```
> 
> then run your bot using the dev script. You can replace npm with your package manager of choice.
> ```sh
> npm run start
> ```
> 

Reviewer 2: ```suggestion
_Note:_ This uses the `run-p` script which is part of [`npm-run-all`](https://github.com/mysticatea/npm-run-all)
_Note 2:_ Please do note that because the processes are started simultaneously you should run `build` at least once before running `dev`, otherwise the `start` process will fail because there are no files to run just yet.
```

    <td align="center"><a href="https://c43721.github.io/"><img src="https://avatars.githubusercontent.com/u/55610086?v=4?s=100" width="100px;" alt=""/><br /><sub><b>c43721</b></sub></a><br /><a href="https://github.com/sapphiredev/plugins/commits?author=c43721" title="Code">💻</a></td>
    <td align="center"><a href="https://megatank58.me/"><img src="https://avatars.githubusercontent.com/u/51410502?v=4?s=100" width="100px;" alt=""/><br /><sub><b>megatank58</b></sub></a><br /><a href="https://github.com/sapphiredev/plugins/commits?author=megatank58" title="Code">💻</a></td>
    <td align="center"><a href="https://fc5570.me/"><img src="https://avatars.githubusercontent.com/u/68158483?v=4?s=100" width="100px;" alt=""/><br /><sub><b>FC</b></sub></a><br /><a href="https://github.com/sapphiredev/plugins/commits?author=FC5570" title="Code">💻</a></td>
  </tr>
</table>

<!-- markdownlint-restore -->
<!-- prettier-ignore-end -->

<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification.

> Contributions of any kind welcome!

Reviewer 2: Dump the content from `/README.md` in here after reverting the changes as requested
